mandible/metadata_mapper/format/format.py

```diff
@@ -73,6 +73,8 @@ def _eval_key_wrapper(self, data, key: Key) -> Any:
         try:
             return self._eval_key(data, key)
         except KeyError as e:
+            if key.optional_key:
```

Comment:
I was originally thinking about having the logic out of the except block, but after some thinking, I don't think that is the route I want to take. If I were to go that route each class that inherits from SimpleFormat would have to have that logic defined in self._eval_key.

Reply:
I do feel like this is better handled by each implementation of the SimpleFormat and using the try except feels wrong to me.

For example each of the systems have a `_eval_key()` defined for example in the jsonpath which json and zip use it seems a get_key() using the jsonpath is set up like a filter. if key.optional_key and not matches return a value.

https://github.com/asfadmin/mandible/blob/86e8c55f6c306057c659e8e7c357f465c4bdd190/mandible/jsonpath.py#L23

for the h5py I believe that groups could be leveraged here to find the keys and check if the optional key is there. https://docs.h5py.org/en/stable/high/group.html rather than the `data["key.key"]` that would lead to a `KeyError`
https://github.com/asfadmin/mandible/blob/86e8c55f6c306057c659e8e7c357f465c4bdd190/mandible/metadata_mapper/format/h5.py#L19

for the xml parse i believe is using the same filter idea from the xpath, returns a match or []. could use if key.optional_key and not matches return a value.
https://github.com/asfadmin/mandible/blob/86e8c55f6c306057c659e8e7c357f465c4bdd190/mandible/metadata_mapper/format/xml.py#L20

Reply:
Generally, I agree. In this case, I don't think I do. If we were to introduce a new type, that is four locations that the logic is being defined. Not to mention, we have to remember to add it to that new type. If we want to change anything thing in the future, that is four cases we need to update.

Reply:
From my understanding, even the Python core dev team will use exceptions for flow control. 
<img width="923" alt="Screenshot 2024-08-08 at 4 34 10 PM" src="https://github.com/user-attachments/assets/221d2c80-e1ba-46b8-9dc3-54a6ad0a6607">

Reply:
It isn't something that I think should be done often, in this case, I think it makes sense.

Reply:
The implementation of a data format needs to define the logic for what it means to handle an optional key.

Reply:
If the general consensus is this is a good way to do it then I will be fine.

Reply:
Assuming there are cases that we need to handle it differently, we can always overwrite the method `_eval_key_wrapper` in that particular data format class. I think returning `None` when the 'optional_key` is set to True and there is no key, will be the solution most implementations need. Going this route, less code needs to be written and maintained.

Reply:
I think this is where I would put it. Feels like a good spot to me. Besides, the addition of this option check doesn't change anything about how the exceptions flow through the program. We need to catch and re-raise anyway because we need to be transforming the exception type.